subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -42,47 +46,53 @@
         this.taskInputsListener = taskInputsListener;
         this.buildOutputCleanupRegistry = buildOutputCleanupRegistry;
-        this.taskOutputChangesListener = taskOutputChangesListener;
+        this.outputChangeListener = outputChangeListener;
         this.executer = executer;
         this.buildInvocationScopeId = buildInvocationScopeId;
     }
 
-    public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
+    public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, final TaskExecutionContext context) {
         TaskProperties taskProperties = context.getTaskProperties();
         FileCollection sourceFiles = taskProperties.getSourceFiles();
         if (taskProperties.hasSourceFiles() && sourceFiles.isEmpty()) {
-            TaskArtifactState taskArtifactState = context.getTaskArtifactState();
-            TaskExecutionHistory executionHistory = taskArtifactState.getExecutionHistory();
-            Set<File> outputFiles = executionHistory.getOutputFiles();
+            AfterPreviousExecutionState previousExecution = context.getAfterPreviousExecution();
+            @SuppressWarnings("RedundantTypeArguments")
+            ImmutableSortedMap<String, FileCollectionFingerprint> outputFiles = previousExecution == null
+                ? ImmutableSortedMap.<String, FileCollectionFingerprint>of()
+                : previousExecution.getOutputFileProperties();
             if (outputFiles.isEmpty()) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no previous output files.", task);
             } else {
-                boolean cleanupDirectories = executionHistory.getOverlappingOutputs() == null;
+                TaskArtifactState taskArtifactState = context.getTaskArtifactState();
+                boolean cleanupDirectories = taskArtifactState.getOverlappingOutputs() == null;
                 if (!cleanupDirectories) {
                     LOGGER.info("No leftover directories for {} will be deleted since overlapping outputs were detected.", task);
                 }
-                taskOutputChangesListener.beforeTaskOutputChanged();
+                outputChangeListener.beforeOutputChange();
                 boolean deletedFiles = false;
                 boolean debugEnabled = LOGGER.isDebugEnabled();
 
-                for (File file : outputFiles) {
-                    if (file.exists() && buildOutputCleanupRegistry.isOutputOwnedByBuild(file)) {
-                        if (!cleanupDirectories && file.isDirectory()) {
-                            continue;
+                for (FileCollectionFingerprint outputFingerprints : outputFiles.values()) {
+                    for (String outputRootPath : outputFingerprints.getFingerprints().keySet()) {
```

Comment:
Why is `root` part of the name `outputRootPath`? This seems to be an `outputPath` or `outputFilePath` anywhere in the output.

Reply:
You're right, this is a leftover. Removed.